### PR TITLE
[JEP-222] Add experimental WebSocket support + Jetty update in Winstone + Remoting 4.0

### PR DESCRIPTION
**Pull request by jglick**
See [JEP-222](https://jenkins.io/jep/222). Downstream of ~https://github.com/jenkinsci/winstone/pull/79~ + ~https://github.com/jenkinsci/remoting/pull/357~ + ~https://github.com/jenkinsci/jenkins-test-harness/pull/183~.

### Proposed Changelog Entries

* Added **experimental** support for connecting inbound agents using a WebSocket transport over HTTP(S) rather than the TCP port, for easier network setup. See [JEP-222](https://jenkins.io/jep/222) for details.
* Remoting 4.0 release, mostly to highlight the binary incompatibility in the last 3.x release: https://github.com/jenkinsci/remoting/releases/tag/remoting-4.0
* Jetty update in Winstone: https://github.com/jenkinsci/winstone/releases/tag/winstone-5.5

### CODE REVIEW

**Comment by jglick** [comment]
Looks like there are some test failures.

File: core/src/main/java/jenkins/slaves/WebSocketAgents.java
Comment (jeffret-b) [inline core/src/main/java/jenkins/slaves/WebSocketAgents.java]: Should the Remoting ping thread be adjusted to be faster in these scenarios? Or turned off in favor of other mechanisms, like this? Duplicate ping mechanisms over a channel can sometimes get in each other's way.

I see some later comments about necessary ping intervals based on different technologies. I'm not sure what makes the most sense here.

File: core/src/main/java/jenkins/websocket/WebSocketEcho.java
Comment (jeffret-b) [inline core/src/main/java/jenkins/websocket/WebSocketEcho.java]: I'm not understanding yet why this is needed.

**Review by jeffret-b — COMMENTED** [review]
So far I've just done a quick overview review. It looks good. It's well structured and easy to follow. I like how nicely it fits into the existing structure while providing valuable new capability.

I'll do a more thorough review at a later time.

File: test-pom/pom.xml
Comment (jglick) [inline test-pom/pom.xml]: note https://github.com/apache/netbeans/pull/1766

File: core/src/main/java/jenkins/slaves/WebSocketAgents.java
Comment (jglick) [reply core/src/main/java/jenkins/slaves/WebSocketAgents.java]: I think it is OK to keep both, at least pending further experience. The Remoting ping is by default every 5m with a 4m timeout, and beyond network connectivity would verify that both sides of the channel are actually processing packets. The WebSocket ping introduced here is unidirectional (server → client), does not wait for pongs unless the underlying library does it automatically, and serves only to prevent certain reverse proxies from closing the connection prematurely.

File: core/src/main/java/jenkins/websocket/WebSocketEcho.java
Comment (jglick) [reply core/src/main/java/jenkins/websocket/WebSocketEcho.java]: This is just here for testing and diagnostics—it is the most convenient way to check whether Jenkins is actually serving WebSocket connections, including through load balancers and whatnot, and to verify keepalive behavior. [JEP-222: _Interactive Tests_](https://github.com/jenkinsci/jep/blob/master/jep/222/README.adoc#interactive-tests)

File: core/src/main/java/hudson/slaves/JNLPLauncher.java
Comment (jtnord) [inline core/src/main/java/hudson/slaves/JNLPLauncher.java]: should this be a boolean.  with kafka for remoting and others I am wondering if it should be an enum or something else.

File: core/src/main/java/jenkins/slaves/WebSocketAgents.java
Comment (jtnord) [inline core/src/main/java/jenkins/slaves/WebSocketAgents.java]: probably we can't but could we put new code in `jenkins.agents`?

File: core/src/main/java/hudson/slaves/JNLPLauncher.java
Comment (jglick) [reply core/src/main/java/hudson/slaves/JNLPLauncher.java]: The Kafka implementation as I understand it is quite different and does not integrate with `JNLPLauncher`.

See also [discussion](https://github.com/jenkinsci/jep/blob/master/jep/222/README.adoc#dedicated-computerlauncher).

File: test/src/test/java/hudson/slaves/JNLPLauncherTest.java
Comment (jtnord) [inline test/src/test/java/hudson/slaves/JNLPLauncherTest.java]: unused in test - is this to show that your fine logging does not blow up due to invalid parameters or other. (in which case a comment in the code would help future reviewers)

File: core/src/main/java/jenkins/slaves/WebSocketAgents.java
Comment (jglick) [reply core/src/main/java/jenkins/slaves/WebSocketAgents.java]: The code base is already split between `hudson.slaves` and `jenkins.slaves`, and various API types already use the term _slave_, so this feels impractical.

File: test/src/test/java/hudson/slaves/JNLPLauncherTest.java
Comment (jglick) [reply test/src/test/java/hudson/slaves/JNLPLauncherTest.java]: Does not affect test pass status, just there to help developers understand what is happening when the test is being run interactively. I.e., the typical usage of `LoggerRule`.

File: cli/pom.xml
Comment (jglick) [inline cli/pom.xml]: https://github.com/jenkinsci/remoting/pull/357#discussion_r360519231

**Comment by jeffret-b** [comment]
I pulled down this PR branch and built it but when I try to run I get
```
Exception in thread "main" java.lang.IllegalAccessException: Can not set static final java.lang.String field winstone.WinstoneSession.SESSION_COOKIE_NAME to java.lang.String
```
Is there anything else I need to update or configure to get this latest version to run?

**Comment by jglick** [comment]
I have not seen such an error. Are you using Java 11?

**Comment by jeffret-b** [comment]
> I have not seen such an error. Are you using Java 11?

No, I'm on Java 8. It worked just fine yesterday. I pulled down the latest this morning and rebuilt and now it isn't working. I don't have a clue why yet.

**Comment by jglick** [comment]
I am on it.

File: core/src/main/java/jenkins/websocket/WebSocketEcho.java
Comment (jeffret-b) [reply core/src/main/java/jenkins/websocket/WebSocketEcho.java]: I'm a little concerned about potential security implications here. Not that I can enumerate any specific ones, though. It's just that this provides an unauthenticated channel. There probably isn't a a DOS attack, because there isn't any noticeable multiplier. Hmmm ... maybe it could be used to pump a lot of worthless, unauthenticated data through a Jenkins server. That sounds like a DOS attack.

**Review by jeffret-b — COMMENTED** [review]
The code looks good. I left a few comments.

I want to test this out before stamping it with an approval. I left a few notes to myself of what to try. I'll try to do the testing tomorrow.

File: cli/src/main/java/hudson/cli/CLI.java
Comment (jglick) [reply cli/src/main/java/hudson/cli/CLI.java]: I just thought it would be convenient—while you can set the Jenkins URL in a profile (or `withEnv`…), you cannot currently do the same for the mode.

File: core/src/main/java/jenkins/websocket/WebSocketEcho.java
Comment (jglick) [reply core/src/main/java/jenkins/websocket/WebSocketEcho.java]: More than just hitting, say, `http://jenkins/login` repeatedly?

File: core/src/main/java/jenkins/slaves/WebSocketAgents.java
Comment (jglick) [reply core/src/main/java/jenkins/slaves/WebSocketAgents.java]: As you prefer.

File: core/src/main/java/hudson/slaves/JNLPLauncher.java
Comment (jglick) [reply core/src/main/java/hudson/slaves/JNLPLauncher.java]: Right, it is mutually exclusive in `remoting.jar`, I am just doing a precheck here so you do not get as far as trying to actually launch the agent and then seeing an error from the CLI.

File: core/src/main/java/jenkins/slaves/WebSocketAgents.java
Comment (jglick) [reply core/src/main/java/jenkins/slaves/WebSocketAgents.java]: There is special handling in the GUI to decide whether or not to show the JNLP button, but I agree nothing is needed here.

File: cli/src/main/java/hudson/cli/CLI.java
Comment (jeffret-b) [reply cli/src/main/java/hudson/cli/CLI.java]: Maybe. If you want to do, I won't object, but I'm just not convinced of the value.

File: core/src/main/java/jenkins/websocket/WebSocketEcho.java
Comment (jeffret-b) [reply core/src/main/java/jenkins/websocket/WebSocketEcho.java]: I don't know. It seems like this allows someone to open a socket and just stream data through and back, which seems to be worse than just hitting the login page. Maybe it should just send back an abbreviated or limited length message. The first 10 characters or something.

File: cli/src/main/java/hudson/cli/CLI.java
Comment (jglick) [reply cli/src/main/java/hudson/cli/CLI.java]: Not intending to do it here; just left as a possibility for the future.

File: core/src/main/java/jenkins/websocket/WebSocketEcho.java
Comment (jglick) [reply core/src/main/java/jenkins/websocket/WebSocketEcho.java]: I cannot see that making any difference; one way or another it needs to handle the incoming message.

Maybe I will just restrict this to administrators. (Unlike TCP connections, it is trivial to pass authentication.)

Comment (jeffret-b) [reply core/src/main/java/jenkins/websocket/WebSocketEcho.java]: True. I do like the idea of restricting it to administrators. That removes concerns.

Comment (jglick) [reply core/src/main/java/jenkins/websocket/WebSocketEcho.java]: [Serendipitously](https://github.com/vi/websocat/releases/tag/v1.5.0)

> Using client URI's like `websocat ws://user:password@host/` now adds basic authentication HTTP header to request

File: core/src/main/resources/hudson/slaves/JNLPLauncher/config.jelly
Comment (jeffret-b) [reply core/src/main/resources/hudson/slaves/JNLPLauncher/config.jelly]: This works a little differently than I expected. Or a little inconsistently.

A) If I don't enable WebSocket here:
  1) I can launch an agent normally and it connects via the TCP port.
  2) I can launch an agent with the "-webSocket" option and it connects via WebSocket.
  3) I can launch an agent with a direct connection, "-direct".

B) If I do enable WebSocket here:
  1) If I try to launch an agent normally, it gets the "-webSocket" argument in the jnlp file and connects via WebSocket.
  2) I can launch an agent with the "-webSocket" option and it connects via WebSocket.
  3) I can launch an agent with a direct connection, "-direct".

In other words, the agent can control if it wants to connect via WebSocket or direct but it can't control if it wants to connect via jnlp file to TCP port.

Things are better when they are regular and consistent. But I'm not certain this is really a problem here, that it needs to be changed, or what it would be changed to. 

Theoretically the config UI could have a choice to connect via i) config (jnlp) file, ii) direct, or iii) WebSocket and then the server would only allow that mechanism. 

Another possibility would be that the WebSocket implementation check connections and not allow them if they're not configured for WebSocket.

The inconsistent behavior had me confused for a while until I mapped it out but maybe it's not going to be confusing for users and developers.

Just recording my observations and musing about it.

Comment (jglick) [reply core/src/main/resources/hudson/slaves/JNLPLauncher/config.jelly]: You are correct that the server does not attempt to reject connections made using a different transport than the selected one. (`-direct` is not a distinct option—this is still using the TCP port, just with the port number being discovered differently.) The checkbox controls whether the CLI argument `-webSocket` is offered to `Main` when `Launcher` is run with `-jnlpUrl`. In other words, the server merely uses this to serve up appropriate `/computer/xxx/slave-agent.jnlp` content.

Comment (jglick) [reply core/src/main/resources/hudson/slaves/JNLPLauncher/config.jelly]: Remember that `Main` is the fundamental entry point for inbound agents. `Launcher` is the wrapper entry point that parses JNLP to call `Main`, as well as handling stdio mode for outbound agents (and a TCP _server_ mode which AFAIK is unused by any Jenkins `ComputerLauncher`). It would be nice if Javadoc and/or Markdown clarified the distinction, as it is hard to keep track of (see https://github.com/jenkinsci/docker-jnlp-slave/issues/87 for example though I think I am abandoning that).

File: core/src/main/java/jenkins/websocket/WebSocketEcho.java
Comment (jeffret-b) [reply core/src/main/java/jenkins/websocket/WebSocketEcho.java]: That's nicely handy and timely. The behavior seems reasonable and works fine now.

File: core/src/main/resources/hudson/slaves/JNLPLauncher/config.jelly
Comment (jeffret-b) [reply core/src/main/resources/hudson/slaves/JNLPLauncher/config.jelly]: Yeah, I think the existing behavior is pretty reasonable the more I think about it. I wanted to record my observations and see if anything else popped out.

File: cli/src/main/java/hudson/cli/CLI.java
Comment (jeffret-b) [reply cli/src/main/java/hudson/cli/CLI.java]: Works fine for the things I tried.

File: core/src/main/resources/hudson/slaves/SlaveComputer/slave-agent.jnlp.jelly
Comment (jeffret-b) [reply core/src/main/resources/hudson/slaves/SlaveComputer/slave-agent.jnlp.jelly]: Works fine for the basic scenarios I've tested.

**Review by jeffret-b — APPROVED** [review]
This looks nice. Everything is in good shape and this will make for a great addition to the agent capabilities.

The code all looks good. I've reviewed the latest changes.

I've tested the capabilities with agents and CLI. Things behave fine. My testing has been mostly happy-path. I've tested out some conflicting or incorrect parameters but I've only tested on a reliable network with straightforward configurations and use. I haven't done any robustness testing.

**Comment by oleg-nenashev** [comment]
@jglick what is the rollout plan for this change? Do you want to have it as `Beta` in the next LTS or just as GA? It might be useful to put it into the JEP explicitly

**Review by res0nance — APPROVED** [review]

**Review by fcojfernandez — APPROVED** [review]
I haven't tested it, but the code looks good

**Comment by jglick** [comment]
> Do you want to have it as `Beta` in the next LTS or just as GA?

I do not think there is any formal mechanism for any such marking. We can phrase it as being experimental in the weekly changelog of course.

**Comment by oleg-nenashev** [comment]
> I do not think there is any formal mechanism for any such marking. We can phrase it as being experimental in the weekly changelog of course.

Changelog would be enough. APIs are already marked as `@Beta`

**Comment by jglick** [comment]
> APIs are already marked as `@Beta`

And could probably stay that way unless plugins start to want to implement their own endpoints.

**Review by timja — APPROVED** [review]
Haven't tested but code looks good

**Comment by oleg-nenashev** [comment]
requesting reviews from the @jenkinsci/core team since it is a major change under the hood of Jenkins. Unfortunately I do not have bandwidth to review it till the next week (wed or so), but I do not mind if it gets merged.with the current approvals.

@jglick Could you please also add the "collateral improvements" from Remoting and Winstone to the changelog?

* Major RFE: Remoting 4.0 release, mostly to highlight the binary incompatible in 4.0 and previous 3.x. https://github.com/jenkinsci/remoting/releases/tag/remoting-4.0
* RFE: Jetty update in Winstone 5.5: https://github.com/jenkinsci/winstone/releases/tag/winstone-5.5

**Comment by oleg-nenashev** [comment]
I plan to merge it tomorrow if no negative feedback

**Comment by jglick** [comment]
> requesting reviews from the @jenkinsci/core team

Note that I already requested reviews from @jenkinsci/core-pr-reviewers.

**Comment by oleg-nenashev** [comment]
@jglick would you be interested to write a blog post about this new feature?

**Comment by jglick** [comment]
Yes I should be able to do that.

**Comment by jglick** [comment]
Merged toward 2.216 FTR.

**Pull request merged**
